modules/core/src/mathfuncs.cpp

```diff
@@ -1563,35 +1565,54 @@ int cv::solveCubic( InputArray _coeffs, OutputArray _roots )
     Mat roots = _roots.getMat();
 
     int i = -1, n = 0;
-    double a0 = 1., a1, a2, a3;
+    // We store raw coefficients to preserve precision for the quadratic case
+    double raw_a0 = 1., raw_a1, raw_a2, raw_a3;
```

Comment:
please, don't introduce raw_a0, raw_a1, raw_a2 and raw_a3. Normalization is quite simple operation that can be done in-place. And it will also further reduce the patch size and will make it easier to analyze.